src/serverlatency.cpp

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "serverlatency.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "pingsenderfactory.h"
+
+#include <QDateTime>
+
+constexpr const uint32_t SERVER_LATENCY_TIMEOUT_MSEC = 5000;
+
+namespace {
+Logger logger(LOG_MAIN, "ServerLatency");
+}
+
+ServerLatency::ServerLatency() { MVPN_COUNT_CTOR(ServerLatency); }
+
+ServerLatency::~ServerLatency() { MVPN_COUNT_DTOR(ServerLatency); }
+
+void ServerLatency::initialize() {
+  MozillaVPN* vpn = MozillaVPN::instance();
+  connect(vpn->controller(), &Controller::stateChanged, this,
+          &ServerLatency::stateChanged);
+}
+
+void ServerLatency::start() {
+  MozillaVPN* vpn = MozillaVPN::instance();
+  if (vpn->controller()->state() != Controller::StateOff) {
+    m_wantRefresh = true;
+    return;
+  }
+
+  m_wantRefresh = false;
+  m_pingSender = PingSenderFactory::create(QHostAddress(), this);
+  ServerCountryModel* scm = vpn->serverCountryModel();
+
+  connect(m_pingSender, SIGNAL(recvPing(quint16)), this,
+          SLOT(recvPing(quint16)));
+  connect(m_pingSender, SIGNAL(criticalPingError()), this,
+          SLOT(criticalPingError()));
+  connect(&m_timeout, &QTimer::timeout, this, &ServerLatency::stop);
+
+  // Send a ping to every server.
+  quint16 seq = 1;
+  for (const Server& server : scm->servers()) {
```

Comment:
I don't like the idea to ping all the servers at the same time. This means hundreds of ICMP packets and that can reach the rate limit on some OS.
I would suggest to use chunks of ICMP packets.

Reply:
Done!